src/Ast/BaseNode.php

```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDocParser\Ast;
+
+abstract class BaseNode implements Node
+{
+
+	/** @var array<string, mixed> */
+	private $attributes = [];
+
+	/**
+	 * @param mixed $value
+	 */
+	public function setAttribute(string $key, $value): void
+	{
+		$this->attributes[$key] = $value;
+	}
+
+	public function hasAttribute(string $key): bool
+	{
+		return array_key_exists($key, $this->attributes);
+	}
+
+	/**
+	 * @param mixed|null $default
+	 * @return mixed|null
+	 */
+	public function getAttribute(string $key, $default = null)
```

Comment:
why do you need the `$default`?

Reply:
Not really, just convention from php-parser. It can be removed

Reply:
Yeah, please always return null if the attribute doesn't exist, I don't think we'd want to override it.

Reply:
I will :+1: